File: backend/app/schemas/tax.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class PersonalInputSchema(BaseModel):
    tax_year: int = Field(2026, ge=2020, le=2030)
    is_married: bool = False
    num_children: int = Field(0, ge=0, le=20)
    is_church_member: bool = False
    church_tax_rate_type: str = Field("high", pattern="^(high|low)$")
    is_full_year_resident: bool = True
    is_disabled: bool = False
    disability_grade: int = Field(0, ge=0, le=100)
    occupation_type: str = Field(
        "employee",
        pattern="^(employee|teacher_civil_servant|freelancer)$",
        description="Occupation type — enables teacher/Beamte-specific deduction suggestions",
    )


class EmploymentInputSchema(BaseModel):
    gross_salary: float = Field(0.0, ge=0, le=500_000)
    lohnsteuer_withheld: float = Field(0.0, ge=0)
    soli_withheld: float = Field(0.0, ge=0)
    kirchensteuer_withheld: float = Field(0.0, ge=0)


class SelfEmployedInputSchema(BaseModel):
    revenue: float = Field(0.0, ge=0)
    expenses: float = Field(0.0, ge=0)


class InvestmentInputSchema(BaseModel):
    gross_income: float = Field(0.0, ge=0)
    tax_withheld: float = Field(0.0, ge=0)
    fund_type: str = Field(
        "standard",
        pattern="^(standard|equity_etf|mixed_fund|real_estate_fund|bond_fund)$",
        description="InvStG fund type — determines Teilfreistellungsquote",
    )
    vorabpauschale: float = Field(
        0.0,
        ge=0,
        description="Vorabpauschale already withheld by broker during the year (§18 InvStG)",
    )


class RentalInputSchema(BaseModel):
    gross_income: float = Field(0.0, ge=0)
    expenses: float = Field(0.0, ge=0)


class DeductionsInputSchema(BaseModel):
    commute_km: float = Field(0.0, ge=0, le=500)
    commute_days: int = Field(0, ge=0, le=250)
    home_office_days: int = Field(0, ge=0, le=365)
    work_equipment: float = Field(0.0, ge=0)
    work_training: float = Field(0.0, ge=0)
    other_work_expenses: float = Field(0.0, ge=0)
    union_fees: float = Field(0.0, ge=0)
    loss_carry_forward: float = Field(
        0.0,
        ge=0,
        description="§10d EStG Verlustvortrag — losses carried forward from prior years",
    )
    # ── Häusliches Arbeitszimmer (§4 Abs.5 Nr.6b / §9 Abs.5 EStG) ──────────
    home_office_type: str = Field(
        "pauschale",
        pattern="^(pauschale|arbeitszimmer)$",
        description="'pauschale' = €6/day flat rate; 'arbeitszimmer' = proportional rent for a dedicated room",
    )
    arbeitszimmer_mittelpunkt: bool = Field(
        False,
        description="True only if the home office is the Mittelpunkt (centre) of all professional activity (>50% of work)",
    )
    apartment_sqm: float = Field(
        0.0, ge=0, le=1_000, description="Total floor area of the apartment in m²"
    )
    office_sqm: float = Field(
        0.0, ge=0, le=500, description="Floor area of the dedicated office room in m²"
    )
    monthly_warm_rent: float = Field(
        0.0,
        ge=0,
        le=50_000,
        description="Total monthly warm rent + Nebenkosten for the whole apartment",
    )
    your_rent_share_pct: float = Field(
        100.0,
        ge=0,
        le=100,
        description="Your percentage share of the rent (100 if you pay alone; 50 if split equally with partner)",
    )
    arbeitszimmer_start_month: int = Field(
        1,
        ge=1,
        le=12,
        description="Month the dedicated office room was first used for work (1=January … 12=December). Prorates the deduction to active months only.",
    )
    # ── Teacher / Civil-servant specific (§9 EStG) ──────────────────────────
    teacher_materials: float = Field(
        0.0,
        ge=0,
        description="Unterrichtsmaterialien: books, worksheets, laminators, stationery (teachers/Beamte only)",
    )
    double_household_costs_per_month: float = Field(
        0.0,
        ge=0,
        le=1_500,
        description="Monthly accommodation costs for the second home near workplace (Doppelte Haushaltsführung, capped at €1,000/month)",
    )
    double_household_months: int = Field(
        0,
        ge=0,
        le=12,
        description="Number of months the double household was maintained in the tax year",
    )


class SpecialExpensesInputSchema(BaseModel):
    health_insurance: float = Field(0.0, ge=0)
    long_term_care_insurance: float = Field(0.0, ge=0)
    pension_contributions: float = Field(0.0, ge=0)
    riester_contributions: float = Field(0.0, ge=0)
    donations: float = Field(0.0, ge=0)
    childcare_costs: float = Field(0.0, ge=0)
    alimony_paid: float = Field(0.0, ge=0)
    church_fees_paid: float = Field(0.0, ge=0)
    medical_costs: float = Field(0.0, ge=0)


class TaxCalculationRequest(BaseModel):
    personal: PersonalInputSchema = PersonalInputSchema()
    employment: EmploymentInputSchema = EmploymentInputSchema()
    self_employed: SelfEmployedInputSchema = SelfEmployedInputSchema()
    investments: InvestmentInputSchema = InvestmentInputSchema()
    rental: RentalInputSchema = RentalInputSchema()
    deductions: DeductionsInputSchema = DeductionsInputSchema()
    special_expenses: SpecialExpensesInputSchema = SpecialExpensesInputSchema()


class TaxBreakdownResponse(BaseModel):
    employment_gross: float
    self_employed_net: float
    investment_income: float
    rental_net: float
    gesamtbetrag_der_einkuenfte: float
    werbungskosten_actual: float
    werbungskosten_pauschale: float
    werbungskosten_used: float
    sonderausgaben_actual: float
    sonderausgaben_pauschale: float
    sonderausgaben_used: float
    aussergewoehnliche_belastungen: float
    disability_pauschbetrag_used: float
    kinderfreibetrag_used: float
    kindergeld_annual: float
    zve: float
    tarifliche_est: float
    solidaritaetszuschlag: float
    kirchensteuer: float
    total_tax: float
    capital_tax_flat: float
    sparer_pauschbetrag_used: float
    teilfreistellung_applied: float
    lohnsteuer_withheld: float
    soli_withheld: float
    kirchensteuer_withheld: float
    capital_tax_withheld: float
    total_withheld: float
    refund_or_payment: float
    effective_rate_percent: float
    marginal_rate_percent: float
    suggestions: list[str]
    tax_year: int


class TaxParametersPublicSchema(BaseModel):
    year: int
    is_active: bool
    notes: Optional[str] = None
    grundfreibetrag: float
    zone2_limit: float
    zone3_limit: float
    zone4_limit: float
    zone2_coeff1: float
    zone2_coeff2: float
    zone3_coeff1: float
    zone3_coeff2: float
    zone3_offset: float
    zone4_rate: float
    zone4_offset: float
    zone5_rate: float
    zone5_offset: float
    kinderfreibetrag: float
    werbungskosten_pauschale: float
    sonderausgaben_pauschale_single: float
    sonderausgaben_pauschale_joint: float
    sparer_pauschbetrag: float
    pendlerpauschale_per_km: float
    homeoffice_per_day: float
    homeoffice_max_days: int
    kindergeld_per_month: float
    soli_rate: float
    soli_freigrenze_single: float
    soli_freigrenze_joint: float
    kirchensteuer_rate_high: float
    kirchensteuer_rate_low: float
    max_pension_deduction_single: float
    max_pension_deduction_joint: float
    alimony_max: float
    ehrenamt_allowance: float
    uebungsleiter_allowance: float
    childcare_rate: float
    childcare_max_per_child: float

    class Config:
        from_attributes = True

File: frontend/src/types/tax.ts
// ─────────────────────────────────────────────────────────────────────────────
//  Wizard State Sections
// ─────────────────────────────────────────────────────────────────────────────
export interface PersonalData {
    isMarried: boolean
    numChildren: number
    isChurchMember: boolean
    churchTaxRateType: 'high' | 'low'  // high = 9% (most states), low = 8% (Bavaria/BW)
    taxYear: number
    federalState?: string
    isFullYearResident?: boolean
    isDisabled?: boolean
    disabilityGrade?: number            // 0 = none, 25–30–50–60–70–80–90–100 = Grad der Behinderung
    occupationType?: 'employee' | 'teacher_civil_servant' | 'freelancer'
}

export interface EmploymentData {
    grossSalary: number
    taxesWithheld: number
    soliWithheld?: number        // Soli Zeile 5 on Lohnsteuerbescheinigung
    kirchensteuerWithheld?: number  // KiSt Zeile 6+7 on Lohnsteuerbescheinigung
    bonus: number             // resolved euro amount always used for calculation
    bonusType?: 'fixed' | 'percent'  // UI preference — fixed € or % of gross salary
    bonusPercent?: number    // percentage value when bonusType === 'percent'
    // Salary-change support: if hasSalaryChange is true, periods override grossSalary
    hasSalaryChange?: boolean
    salaryPeriods?: SalaryPeriod[]
}

export interface SalaryPeriod {
    months: number      // number of months at this salary (1–12)
    monthlyGross: number  // gross monthly salary for this period
}

export interface OtherIncomeData {
    selfEmployedRevenue: number
    selfEmployedExpenses: number
    dividends: number
    capitalGains: number
    capitalTaxesWithheld: number
    rentalIncome: number
    rentalExpenses: number
    // ETF / fund type — drives Teilfreistellung (InvStG 2018)
    fundType?: 'standard' | 'equity_etf' | 'mixed_fund' | 'real_estate_fund' | 'bond_fund'
    vorabpauschale?: number  // §18 InvStG advance lump-sum already withheld by broker
}

export interface DeductionsData {
    commuteKm: number
    commuteDays: number
    homeOfficeDays: number
    otherWorkExpenses: number
    workEquipment?: number   // desk, laptop, chair — fully deductible (GWG ≤€952)
    workTraining?: number    // courses, books, professional education
    unionFees?: number       // Gewerkschaftsbeiträge
    lossCarryForward?: number  // §10d EStG — Verlustvortrag from prior years
    // Häusliches Arbeitszimmer (§4 Abs.5 Nr.6b / §9 Abs.5 EStG)
    homeOfficeType?: 'pauschale' | 'arbeitszimmer'
    arbeitszimmerMittelpunkt?: boolean   // is home the Mittelpunkt of all professional activity?
    apartmentSqm?: number                // total apartment floor area in m²
    officeSqm?: number                   // dedicated office room floor area in m²
    monthlyWarmRent?: number             // total monthly warm rent (€)
    yourRentSharePct?: number            // your % share of rent (0–100; 100 if you pay alone)
    arbeitszimmerStartMonth?: number     // month room was first used for work (1=Jan … 12=Dec); prorates the deduction
    // Teacher / civil-servant specific
    teacherMaterials?: number            // Unterrichtsmaterialien: books, worksheets, etc.
    doubleHouseholdCostsPerMonth?: number  // Doppelte Haushaltsführung monthly costs (€1,000/month cap)
    doubleHouseholdMonths?: number         // number of months maintained
}

export interface SpecialExpensesData {
    pensionContributions: number
    healthInsuranceContributions: number  // GKV/PKV premiums (§10 Abs.1 Nr.3 EStG)
    longTermCareInsurance?: number        // Pflegeversicherung
    riesterContributions: number
    donations: number
    alimonyPaid: number
    churchTaxPriorYear: number
    childcareCosts: number
    medicalCosts?: number                 // Außergewöhnliche Belastungen (§33 EStG)
}

/** Matches the API snake_case response */
export interface TaxYearParameters {
    id: number
    year: number
    is_active: boolean
    grundfreibetrag: number
    zone2_limit: number
    zone3_limit: number
    zone4_limit: number
    zone2_coeff1: number
    zone2_coeff2: number
    zone3_coeff1: number
    zone3_coeff2: number
    zone3_offset: number
    zone4_rate: number
    zone4_offset: number
    zone5_rate: number
    zone5_offset: number
    soli_rate: number
    soli_freigrenze_single: number
    soli_freigrenze_joint: number
    werbungskosten_pauschale: number
    sonderausgaben_pauschale_single: number
    sonderausgaben_pauschale_joint: number
    sparer_pauschbetrag: number
    pendlerpauschale_per_km: number
    homeoffice_per_day: number
    homeoffice_max_days: number
    kinderfreibetrag: number
    kindergeld_per_month: number
    kirchensteuer_rate_high: number
    kirchensteuer_rate_low: number
    max_pension_deduction_single: number
    max_pension_deduction_joint: number
    alimony_max: number
    ehrenamt_allowance: number
    uebungsleiter_allowance: number
    childcare_rate: number
    childcare_max_per_child: number
    notes?: string
}

// ─────────────────────────────────────────────────────────────────────────────
//  Local calculation result (returned by taxCalculator.ts)
// ─────────────────────────────────────────────────────────────────────────────
export interface TaxBreakdown {
    tax_year: number

    // Income breakdown
    gross_income: number
    employment_gross?: number
    self_employed_net?: number
    investment_income?: number
    rental_net?: number
    gesamtbetrag_der_einkuenfte?: number

    // Deductions applied
    werbungskosten_actual?: number
    werbungskosten_pauschale?: number
    werbungskosten_used: number
    sonderausgaben_actual?: number
    sonderausgaben_pauschale?: number
    sonderausgaben_used: number
    aussergewoehnliche_belastungen?: number  // renamed from aussergewoehnliche_belast
    disability_pauschbetrag_used?: number     // §33b EStG flat-rate disability allowance

    // Core tax result
    zve: number
    tarifliche_est: number
    solidaritaetszuschlag: number
    kirchensteuer: number
    kinderfreibetrag_used: number
    kindergeld_annual: number
    capital_tax_flat: number
    capital_tax_due: number
    sparer_pauschbetrag_used?: number
    teilfreistellung_applied?: number  // InvStG exempt amount (equity ETF 30%, mixed 15%, etc.)
    total_tax: number

    // Withheld
    lohnsteuer_withheld?: number
    soli_withheld?: number
    kirchensteuer_withheld?: number
    capital_tax_withheld?: number
    total_withheld: number

    // Bottom line
    refund_or_payment: number

    // Rates  (0–1 decimal scale, NOT percent)
    effective_rate: number
    marginal_rate: number

    suggestions: string[]
}
